File: src/client/router.rs
use anyhow::{Result, anyhow};

use crate::client::event::{
    Authenticate, Cap, Ping, PrivateMessage, RawMessage, SaslSuccess, Welcome,
};

pub(super) trait EventDispatcher {
    fn dispatch<T: 'static>(&mut self, event: T);
}

pub(super) fn dispatch_client_event<T, D>(message: T, dispatcher: &mut D)
where
    T: 'static,
    D: EventDispatcher,
{
    dispatcher.dispatch(message);
}

pub(super) fn dispatch_server_event<D>(raw: RawMessage, dispatcher: &mut D) -> Result<()>
where
    D: EventDispatcher,
{
    match raw.command() {
        "CAP" => process::<Cap, _>(raw, dispatcher),
        "AUTHENTICATE" => process::<Authenticate, _>(raw, dispatcher),
        "903" => process::<SaslSuccess, _>(raw, dispatcher),
        "001" => process::<Welcome, _>(raw, dispatcher),
        "PING" => process::<Ping, _>(raw, dispatcher),
        "PRIVMSG" => process::<PrivateMessage, _>(raw, dispatcher),
        _ => process::<RawMessage, _>(raw, dispatcher),
    }
}

fn process<M, D>(raw: RawMessage, dispatcher: &mut D) -> Result<()>
where
    M: TryFrom<RawMessage> + 'static,
    M::Error: std::fmt::Display,
    D: EventDispatcher,
{
    let message = M::try_from(raw).map_err(|e| anyhow!("failed to convert raw message: {e}"))?;
    dispatcher.dispatch(message);

    Ok(())
}

File: src/client/router.ts
import type { Logger } from 'pino'
import {
  constructAuthenticate,
  constructCap,
  constructPing,
  constructPrivateMessage,
  constructSaslSuccess,
  constructWelcome,
  type Events,
} from './events.js'
import type { Connection } from './index.js'
import { parseRawMessage, type RawMessage } from './message.js'

type EventConstructor<Event extends keyof Events> = (raw: RawMessage) => Events[Event]

export function routeEvent(logger: Logger, connection: Connection, message: string) {
  try {
    const raw = parseRawMessage(message)
    switch (raw.command) {
      case 'PRIVMSG':
        constructAndEmit(connection, raw, constructPrivateMessage)
        break
      case 'PING':
        constructAndEmit(connection, raw, constructPing)
        break
      case 'CAP':
        constructAndEmit(connection, raw, constructCap)
        break
      case 'AUTHENTICATE':
        constructAndEmit(connection, raw, constructAuthenticate)
        break
      case '903':
        constructAndEmit(connection, raw, constructSaslSuccess)
        break
      case '001':
        constructAndEmit(connection, raw, constructWelcome)
        break
    }
  } catch (error) {
    logger.error({ err: error }, 'failed to route event')
  }
}

function constructAndEmit<Event extends keyof Events>(
  connection: Connection,
  raw: RawMessage,
  constructor: EventConstructor<Event>
) {
  connection.emit(constructor(raw))
}
